primitives/xcm/src/barriers.rs

```diff
@@ -77,3 +77,97 @@ impl<T: Contains<MultiLocation>> ShouldExecute for AllowTopLevelPaidExecutionDes
 		}
 	}
 }
+
+/// Make sure that not withdrawable assets are handle properly according to the XCM design:
+/// - forbid any ReserveAssetDeposited instruction that contains not withdrawable asset(s) as
+/// parameter.
+/// - Morph some specific messages patterns that try to Widraw then Deposit not withdrawable
+/// asset(s).
+pub struct NotWithdrawableAssetsBarrier<IsWithdrawable, InnerBarrier>(
+	core::marker::PhantomData<(IsWithdrawable, InnerBarrier)>,
+);
+
+impl<IsWithdrawable: crate::IsWithdrawable, InnerBarrier: ShouldExecute> ShouldExecute
+	for NotWithdrawableAssetsBarrier<IsWithdrawable, InnerBarrier>
+{
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		instructions: &mut Xcm<Call>,
+		max_weight: Weight,
+		weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		InnerBarrier::should_execute(origin, instructions, max_weight, weight_credit)?;
+		let maybe_morphed_message = match &instructions.0[..] {
+			&[WithdrawAsset(ref assets), ClearOrigin, BuyExecution {
```

Comment:
@librelois can you include some documentation (in the code) about why and what is this part of the code.